File: implementations/go/provekit-emit-go-testing/emitter.go
package emitgotesting

import (
	"fmt"
	"go/token"
	"sort"
	"strconv"
	"strings"
	"unicode"

	"github.com/tsavo/provekit/go/provekit-ir-symbolic/canonicalizer"
)

type EmitPlan struct {
	ContractID  string           `json:"contract_id"`
	PackageName string           `json:"package_name"`
	Function    string           `json:"function"`
	Params      []string         `json:"params"`
	ParamTypes  []string         `json:"param_types"`
	ReturnType  string           `json:"return_type"`
	Predicates  []map[string]any `json:"predicates"`
}

type Emission struct {
	Kind                  string   `json:"kind"`
	Source                string   `json:"source"`
	Path                  string   `json:"path"`
	Extension             string   `json:"extension"`
	EmittedArtifactCID    string   `json:"emitted_artifact_cid"`
	EmittedPredicates     []string `json:"emitted_predicates"`
	UnsupportedPredicates []string `json:"unsupported_predicates"`
	IsComplete            bool     `json:"is_complete"`
}

func EmitPlanFromParams(params map[string]any) EmitPlan {
	return EmitPlan{
		ContractID:  firstString(params["contract_id"], params["concept_name"]),
		PackageName: firstString(params["package_name"], params["package"], params["packageName"]),
		Function:    firstString(params["function"], params["function_name"], params["functionName"]),
		Params:      stringList(params["params"]),
		ParamTypes:  stringList(params["param_types"]),
		ReturnType:  firstString(params["return_type"], params["returnType"]),
		Predicates:  mapList(params["predicates"]),
	}
}

func Emit(plan EmitPlan) Emission {
	pkg := sanitizeIdentifier(plan.PackageName, "provekit")
	var functions []string
	var emitted []string
	var unsupported []string
	needsErrors := false

	for i, predicate := range plan.Predicates {
		head := headOf(predicate)
		assertion, ok := renderAssertion(head, predicate)
		if !ok {
			unsupported = append(unsupported, unsupportedName(head))
			continue
		}
		vars := freeVars(predicate)
		decls, declNeedsErrors := declarationsFor(head, vars)
		if declNeedsErrors {
			needsErrors = true
		}
		functions = append(functions, renderTestFunction(head, i, decls, assertion))
		emitted = append(emitted, canonicalHead(head))
	}

	source := renderModule(pkg, functions, needsErrors)
	return Emission{
		Kind:                  "go-testing-test-emission",
		Source:                source,
		Path:                  "provekit_emitted_test.go",
		Extension:             "go",
		EmittedArtifactCID:    canonicalizer.ComputeCID([]byte(source)),
		EmittedPredicates:     emitted,
		UnsupportedPredicates: unsupported,
		IsComplete:            len(unsupported) == 0 && len(emitted) > 0,
	}
}

func renderModule(pkg string, functions []string, needsErrors bool) string {
	var b strings.Builder
	fmt.Fprintf(&b, "package %s\n", pkg)
	if len(functions) > 0 {
		b.WriteByte('\n')
		if needsErrors {
			b.WriteString("import (\n\t\"errors\"\n\t\"testing\"\n)\n")
		} else {
			b.WriteString("import \"testing\"\n")
		}
	}
	for _, fn := range functions {
		b.WriteByte('\n')
		b.WriteString(fn)
		if !strings.HasSuffix(fn, "\n") {
			b.WriteByte('\n')
		}
	}
	return b.String()
}

func renderTestFunction(head string, index int, decls []string, assertion string) string {
	var b strings.Builder
	fmt.Fprintf(&b, "func TestProvekit%s%d(t *testing.T) {\n", camelHead(head), index)
	for _, decl := range decls {
		fmt.Fprintf(&b, "\t%s\n", decl)
	}
	fmt.Fprintf(&b, "\t%s\n", assertion)
	b.WriteString("}\n")
	return b.String()
}

func renderAssertion(head string, predicate map[string]any) (string, bool) {
	args := argsOf(predicate)
	switch canonicalHead(head) {
	case "eq":
		return binaryAssertion(args, "==", "eq")
	case "ne":
		return binaryAssertion(args, "!=", "ne")
	case "lt":
		return binaryAssertion(args, "<", "lt")
	case "gt":
		return binaryAssertion(args, ">", "gt")
	case "le":
		return binaryAssertion(args, "<=", "le")
	case "ge":
		return binaryAssertion(args, ">=", "ge")
	case "option-is-some", "not-null":
		return unaryAssertion(args, "== nil", "option-is-some")
	case "option-is-none":
		return unaryAssertion(args, "!= nil", "option-is-none")
	case "fallible-err":
		if len(args) != 1 {
			return "", false
		}
		x, ok := renderTerm(args[0])
		if !ok {
			return "", false
		}
		return fmt.Sprintf("if %s == nil {\n\t\tt.Fatalf(\"provekit predicate fallible-err failed: expected non-nil error\")\n\t}", x), true
	default:
		return "", false
	}
}

func binaryAssertion(args []map[string]any, op, head string) (string, bool) {
	if len(args) != 2 {
		return "", false
	}
	a, ok := renderTerm(args[0])
	if !ok {
		return "", false
	}
	b, ok := renderTerm(args[1])
	if !ok {
		return "", false
	}
	return fmt.Sprintf("if !(%s %s %s) {\n\t\tt.Fatalf(\"provekit predicate %s failed: %s %s %s\")\n\t}", a, op, b, head, a, op, b), true
}

func unaryAssertion(args []map[string]any, condition, head string) (string, bool) {
	if len(args) != 1 {
		return "", false
	}
	x, ok := renderTerm(args[0])
	if !ok {
		return "", false
	}
	return fmt.Sprintf("if %s %s {\n\t\tt.Fatalf(\"provekit predicate %s failed\")\n\t}", x, condition, head), true
}

func declarationsFor(head string, vars []string) ([]string, bool) {
	decls := make([]string, 0, len(vars))
	needsErrors := false
	for i, name := range vars {
		ident := sanitizeIdentifier(name, fmt.Sprintf("v%d", i))
		switch canonicalHead(head) {
		case "option-is-none":
			decls = append(decls, fmt.Sprintf("var %s any = nil", ident))
		case "option-is-some", "not-null":
			decls = append(decls, fmt.Sprintf("%s := any(1)", ident))
		case "fallible-err":
			needsErrors = true
			decls = append(decls, fmt.Sprintf("%s := errors.New(\"contract error\")", ident))
		default:
			decls = append(decls, fmt.Sprintf("%s := %s", ident, placeholderValue(head, i)))
		}
	}
	return decls, needsErrors
}

func placeholderValue(head string, index int) string {
	switch canonicalHead(head) {
	case "lt":
		if index == 0 {
			return "0"
		}
		return "1"
	case "gt":
		if index == 0 {
			return "1"
		}
		return "0"
	case "ne":
		if index == 0 {
			return "0"
		}
		return "1"
	default:
		return "0"
	}
}

func renderTerm(term map[string]any) (string, bool) {
	kind, _ := term["kind"].(string)
	switch kind {
	case "var":
		name, _ := term["name"].(string)
		ident := sanitizeIdentifier(name, "")
		if ident == "" {
			return "", false
		}
		return ident, true
	case "const":
		return renderConst(term["value"])
	case "op", "ctor":
		return renderApplication(term)
	default:
		return "", false
	}
}

func renderConst(value any) (string, bool) {
	switch v := value.(type) {
	case nil:
		return "nil", true
	case bool:
		if v {
			return "true", true
		}
		return "false", true
	case int:
		return strconv.Itoa(v), true
	case int64:
		return strconv.FormatInt(v, 10), true
	case float64:
		if v == float64(int64(v)) {
			return strconv.FormatInt(int64(v), 10), true
		}
		return strconv.FormatFloat(v, 'g', -1, 64), true
	case string:
		return strconv.Quote(v), true
	default:
		return "", false
	}
}

func renderApplication(term map[string]any) (string, bool) {
	name, _ := term["name"].(string)
	name = strings.TrimPrefix(name, "concept:")
	args := argsOf(term)
	rendered := make([]string, 0, len(args))
	for _, arg := range args {
		r, ok := renderTerm(arg)
		if !ok {
			return "", false
		}
		rendered = append(rendered, r)
	}
	if isArithmetic(name) && len(rendered) == 2 {
		return fmt.Sprintf("(%s %s %s)", rendered[0], name, rendered[1]), true
	}
	ident := sanitizeIdentifier(name, "")
	if ident == "" {
		return "", false
	}
	return fmt.Sprintf("%s(%s)", ident, strings.Join(rendered, ", ")), true
}

func freeVars(term map[string]any) []string {
	var out []string
	seen := map[string]bool{}
	var walk func(map[string]any)
	walk = func(node map[string]any) {
		if node["kind"] == "var" {
			name, _ := node["name"].(string)
			if name != "" && !seen[name] {
				seen[name] = true
				out = append(out, name)
			}
			return
		}
		for _, arg := range argsOf(node) {
			walk(arg)
		}
	}
	walk(term)
	return out
}

func headOf(predicate map[string]any) string {
	name, _ := predicate["name"].(string)
	return strings.TrimPrefix(name, "concept:")
}

func canonicalHead(head string) string {
	switch head {
	case "neq":
		return "ne"
	case "lte":
		return "le"
	case "gte":
		return "ge"
	default:
		return head
	}
}

func unsupportedName(head string) string {
	if head == "" {
		return "<malformed>"
	}
	return head
}

func argsOf(node map[string]any) []map[string]any {
	switch raw := node["args"].(type) {
	case []map[string]any:
		return raw
	case []any:
		out := make([]map[string]any, 0, len(raw))
		for _, item := range raw {
			if m, ok := item.(map[string]any); ok {
				out = append(out, m)
			}
		}
		return out
	default:
		return nil
	}
}

func isArithmetic(name string) bool {
	switch name {
	case "+", "-", "*", "/", "%":
		return true
	default:
		return false
	}
}

func camelHead(head string) string {
	head = canonicalHead(head)
	parts := strings.FieldsFunc(head, func(r rune) bool {
		return r == '-' || r == '_' || r == ':' || r == '.'
	})
	if len(parts) == 0 {
		return "Predicate"
	}
	var b strings.Builder
	for _, part := range parts {
		if part == "" {
			continue
		}
		runes := []rune(part)
		runes[0] = unicode.ToUpper(runes[0])
		b.WriteString(string(runes))
	}
	if b.Len() == 0 {
		return "Predicate"
	}
	return b.String()
}

func sanitizeIdentifier(name, fallback string) string {
	if token.IsIdentifier(name) && !isKeyword(name) {
		return name
	}
	if fallback != "" && token.IsIdentifier(fallback) && !isKeyword(fallback) {
		return fallback
	}
	return ""
}

func isKeyword(name string) bool {
	return token.Lookup(name).IsKeyword()
}

func firstString(values ...any) string {
	for _, value := range values {
		if s, ok := value.(string); ok && strings.TrimSpace(s) != "" {
			return s
		}
	}
	return ""
}

func stringList(value any) []string {
	switch raw := value.(type) {
	case []string:
		return append([]string(nil), raw...)
	case []any:
		out := make([]string, 0, len(raw))
		for _, item := range raw {
			if s, ok := item.(string); ok {
				out = append(out, s)
			}
		}
		return out
	default:
		return nil
	}
}

func mapList(value any) []map[string]any {
	switch raw := value.(type) {
	case []map[string]any:
		return append([]map[string]any(nil), raw...)
	case []any:
		out := make([]map[string]any, 0, len(raw))
		for _, item := range raw {
			if m, ok := item.(map[string]any); ok {
				out = append(out, m)
			}
		}
		return out
	default:
		return nil
	}
}

func supportedPredicates() []string {
	out := []string{
		"concept:eq",
		"concept:ne",
		"concept:lt",
		"concept:gt",
		"concept:le",
		"concept:ge",
		"concept:option-is-some",
		"concept:option-is-none",
		"concept:not-null",
		"concept:fallible-err",
	}
	sort.Strings(out)
	return out
}

File: implementations/typescript/provekit-emit-typescript-vitest/src/emitter.js
"use strict";

const { blake3 } = require("@noble/hashes/blake3.js");

function emit(plan) {
  const normalized = normalizePlan(plan);
  const emittedPredicates = [];
  const unsupportedPredicates = [];
  const tests = [];

  normalized.predicates.forEach((predicate, index) => {
    const head = headOf(predicate);
    const assertion = renderAssertion(head, predicate);
    if (assertion === null) {
      unsupportedPredicates.push(head || "<malformed>");
      return;
    }
    emittedPredicates.push(canonicalHead(head));
    tests.push(renderTest(head, index, declarationsFor(head, freeVars(predicate)), assertion));
  });

  const source = renderModule(normalized.functionName, tests);
  return {
    kind: "typescript-vitest-test-emission",
    source,
    path: modulePath(normalized.functionName),
    extension: "ts",
    emitted_artifact_cid: blake3Cid(source),
    emitted_predicates: emittedPredicates,
    unsupported_predicates: unsupportedPredicates,
    is_complete: unsupportedPredicates.length === 0 && emittedPredicates.length > 0,
  };
}

function normalizePlan(plan) {
  const params = isObject(plan) ? plan : {};
  return {
    contractId: firstString(params.contract_id, params.concept_name),
    functionName: firstString(params.function, params.function_name, params.functionName) || "contract",
    predicates: array(params.predicates).filter(isObject),
  };
}

function renderModule(functionName, tests) {
  const suite = stringLiteral(`provekit contract ${functionName || "contract"}`);
  const body = tests.length > 0 ? tests.join("\n\n") : "  it(\"has no emitted predicates\", () => {});";
  return `describe(${suite}, () => {\n${body}\n});\n`;
}

function renderTest(head, index, declarations, assertion) {
  const testName = stringLiteral(`verifies ${canonicalHead(head)} ${index}`);
  const lines = [`  it(${testName}, () => {`];
  declarations.forEach((decl) => lines.push(`    ${decl}`));
  assertion.split("\n").forEach((line) => lines.push(`    ${line}`));
  lines.push("  });");
  return lines.join("\n");
}

function renderAssertion(head, predicate) {
  const args = array(predicate.args).filter(isObject);
  switch (canonicalHead(head)) {
    case "eq":
      return binaryAssertion(args, "toEqual");
    case "ne":
      return binaryAssertion(args, "not.toEqual");
    case "lt":
      return binaryComparator(args, "<");
    case "gt":
      return binaryComparator(args, ">");
    case "le":
      return binaryComparator(args, "<=");
    case "ge":
      return binaryComparator(args, ">=");
    case "option-is-some":
    case "not-null":
      return unaryNotNull(args);
    case "option-is-none":
      return unaryNull(args);
    default:
      return null;
  }
}

function binaryAssertion(args, matcher) {
  if (args.length !== 2) return null;
  const left = renderTerm(args[0]);
  const right = renderTerm(args[1]);
  if (left === null || right === null) return null;
  return `expect(${left}).${matcher}(${right});`;
}

function binaryComparator(args, op) {
  if (args.length !== 2) return null;
  const left = renderTerm(args[0]);
  const right = renderTerm(args[1]);
  if (left === null || right === null) return null;
  return `expect(${left} ${op} ${right}).toBe(true);`;
}

function unaryNotNull(args) {
  if (args.length !== 1) return null;
  const value = renderTerm(args[0]);
  if (value === null) return null;
  return `expect(${value}).not.toBeNull();`;
}

function unaryNull(args) {
  if (args.length !== 1) return null;
  const value = renderTerm(args[0]);
  if (value === null) return null;
  return `expect(${value}).toBeNull();`;
}

function declarationsFor(head, vars) {
  return vars.map((name, index) => {
    const ident = sanitizeIdentifier(name, `v${index}`);
    return `const ${ident} = ${placeholderValue(head, index)};`;
  });
}

function placeholderValue(head, index) {
  switch (canonicalHead(head)) {
    case "lt":
      return index === 0 ? "0" : "1";
    case "gt":
      return index === 0 ? "1" : "0";
    case "le":
    case "ge":
    case "eq":
      return "0";
    case "ne":
      return index === 0 ? "0" : "1";
    case "option-is-none":
      return "null";
    case "option-is-some":
    case "not-null":
      return "1";
    default:
      return "0";
  }
}

function renderTerm(term) {
  if (!isObject(term)) return null;
  const kind = firstString(term.kind);
  if (kind === "var") {
    return sanitizeIdentifier(firstString(term.name), "");
  }
  if (kind === "const") {
    return renderConst(term.value);
  }
  if (kind === "op" || kind === "ctor") {
    return renderApplication(term);
  }
  return null;
}

function renderConst(value) {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number" && Number.isFinite(value)) return JSON.stringify(value);
  if (typeof value === "string") return JSON.stringify(value);
  return null;
}

function renderApplication(term) {
  const name = firstString(term.name, term.conceptName, term.concept_name);
  const args = array(term.args).filter(isObject);
  const rendered = args.map(renderTerm);
  if (rendered.some((value) => value === null)) return null;
  if (["+", "-", "*", "/", "%"].includes(name) && rendered.length === 2) {
    return `(${rendered[0]} ${name} ${rendered[1]})`;
  }
  if (name === "array") return `[${rendered.join(", ")}]`;
  if (name === "tuple") return `[${rendered.join(", ")}]`;
  const callee = sanitizeIdentifier(name.replace(/^concept:/, ""), "");
  if (callee === "") return null;
  return `${callee}(${rendered.join(", ")})`;
}

function freeVars(predicate) {
  const out = [];
  collectFreeVars(predicate, out);
  return out;
}

function collectFreeVars(value, out) {
  if (Array.isArray(value)) {
    value.forEach((item) => collectFreeVars(item, out));
    return;
  }
  if (!isObject(value)) return;
  if (value.kind === "var") {
    const name = firstString(value.name);
    if (name !== "" && !out.includes(name)) out.push(name);
  }
  collectFreeVars(value.args, out);
}

function headOf(predicate) {
  if (!isObject(predicate)) return "";
  return firstString(predicate.name, predicate.concept_name, predicate.conceptName);
}

function canonicalHead(head) {
  const normalized = firstString(head).replace(/^concept:/, "").replace(/_/g, "-");
  switch (normalized) {
    case "=":
      return "eq";
    case "!=":
    case "\u2260":
      return "ne";
    case "<":
      return "lt";
    case ">":
      return "gt";
    case "<=":
    case "\u2264":
      return "le";
    case ">=":
    case "\u2265":
      return "ge";
    default:
      return normalized;
  }
}

function modulePath(functionName) {
  const safe = firstString(functionName).replace(/[^0-9A-Za-z_]+/g, "_").replace(/^_+|_+$/g, "").toLowerCase();
  return `provekit_${safe || "contract"}.test.ts`;
}

function sanitizeIdentifier(value, fallback) {
  const raw = firstString(value).replace(/-/g, "_").replace(/[^0-9A-Za-z_$]/g, "_");
  const candidate = raw.replace(/^[^A-Za-z_$]+/, "");
  return /^[A-Za-z_$][0-9A-Za-z_$]*$/.test(candidate) ? candidate : fallback;
}

function stringLiteral(value) {
  return JSON.stringify(String(value));
}

function blake3Cid(source) {
  const digest = blake3(Buffer.from(source, "utf8"), { dkLen: 64 });
  return `blake3-512:${Buffer.from(digest).toString("hex")}`;
}

function firstString(...values) {
  for (const value of values) {
    if (typeof value === "string" && value.trim() !== "") return value.trim();
  }
  return "";
}

function array(value) {
  return Array.isArray(value) ? value : [];
}

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

module.exports = {
  emit,
  normalizePlan,
};
